### PR TITLE
[EI_TOOL-71]Incorporate cache mediator changes

### PR DESCRIPTION
**Pull request by prabushi**
Fix for https://github.com/wso2/devstudio-tooling-ei/issues/71

### CODE REVIEW

File: plugins/org.wso2.developerstudio.eclipse.gmf.esb.diagram/src/org/wso2/developerstudio/eclipse/gmf/esb/diagram/custom/deserializer/CacheMediatorExtFactory.java [hunk 1]
```diff
@@ -23,26 +23,26 @@
...
 import org.apache.axiom.om.OMElement;
 import org.apache.synapse.Mediator;
-import org.wso2.carbon.mediator.cache.config.xml.CacheMediatorFactory;
+import org.wso2.carbon.mediator.cache.CacheMediatorFactory;
 import org.wso2.carbon.mediator.cache.digest.DigestGenerator;
 import org.apache.synapse.config.xml.SequenceMediatorFactory;
 import org.apache.synapse.config.xml.XMLConfigConstants;
 import org.wso2.carbon.mediator.cache.CacheMediator;
+//import org.wso2.carbon.mediator.cache.CacheStore;
 import org.wso2.caching.CachingConstants;
```
Comment (manoramahp) [inline plugins/org.wso2.developerstudio.eclipse.gmf.esb.diagram/src/org/wso2/developerstudio/eclipse/gmf/esb/diagram/custom/deserializer/CacheMediatorExtFactory.java]: Shall we remove this commented line?

File: plugins/org.wso2.developerstudio.eclipse.gmf.esb.diagram/src/org/wso2/developerstudio/eclipse/gmf/esb/diagram/custom/deserializer/CacheMediatorExtFactory.java [hunk 2]
```diff
@@ -77,39 +64,60 @@ public Mediator createSpecificMediator(OMElement elem, Properties properties) {
...
+                	cacheMediator.setHeadersToExcludeInHash(child.getText().split(","));
+                }
+                for(Iterator itrChildren = implElem.getChildrenWithLocalName("responseCodes"); itrChildren.hasNext();) {
+                	OMElement child = (OMElement)itrChildren.next();
+                	cacheMediator.setResponseCodes(child.getText());
+                }
+                for(Iterator itrChildren = implElem.getChildrenWithLocalName("hashGenerator"); itrChildren.hasNext();) {
+                	OMElement child = (OMElement)itrChildren.next();
+
+        			if (child.getText() != null && child.getText() != null) {
```
Comment (erandacr) [inline plugins/org.wso2.developerstudio.eclipse.gmf.esb.diagram/src/org/wso2/developerstudio/eclipse/gmf/esb/diagram/custom/deserializer/CacheMediatorExtFactory.java]: Formatting issues (indentation)?

File: plugins/org.wso2.developerstudio.eclipse.gmf.esb/src/org/wso2/developerstudio/eclipse/gmf/esb/CacheProtocolType.java [hunk 3]
```diff
@@ -0,0 +1,200 @@
+/**
+ * Copyright 2009-2012 WSO2, Inc. (http://wso2.com)
+ *
```
Comment (chanikag) [inline plugins/org.wso2.developerstudio.eclipse.gmf.esb/src/org/wso2/developerstudio/eclipse/gmf/esb/CacheProtocolType.java]: wrong license header

File: plugins/org.wso2.developerstudio.eclipse.gmf.esb/src/org/wso2/developerstudio/eclipse/gmf/esb/CacheMediator.java [hunk 4]
```diff
@@ -17,19 +17,21 @@
  * </p>
  * <ul>
  *   <li>{@link org.wso2.developerstudio.eclipse.gmf.esb.CacheMediator#getCacheId <em>Cache Id</em>}</li>
- *   <li>{@link org.wso2.developerstudio.eclipse.gmf.esb.CacheMediator#getCacheScope <em>Cache Scope</em>}</li>
+ *   <li>{@link org.wso2.developerstudio.eclipse.gmf.esb.CacheMediator#getCacheProtocolType <em>Cache Protocol Type</em>}</li>
  *   <li>{@link org.wso2.developerstudio.eclipse.gmf.esb.CacheMediator#getCacheAction <em>Cache Action</em>}</li>
  *   <li>{@link org.wso2.developerstudio.eclipse.gmf.esb.CacheMediator#getHashGenerator <em>Hash Generator</em>}</li>
  *   <li>{@link org.wso2.developerstudio.eclipse.gmf.esb.CacheMediator#getCacheTimeout <em>Cache Timeout</em>}</li>
  *   <li>{@link org.wso2.developerstudio.eclipse.gmf.esb.CacheMediator#getMaxMessageSize <em>Max Message Size</em>}</li>
```
Comment (bsenduran) [inline plugins/org.wso2.developerstudio.eclipse.gmf.esb/src/org/wso2/developerstudio/eclipse/gmf/esb/CacheMediator.java]: Shall we add the default unit within brackets

File: plugins/org.wso2.developerstudio.eclipse.gmf.esb.diagram/src/org/wso2/developerstudio/eclipse/gmf/esb/diagram/custom/deserializer/CacheMediatorDeserializer.java [hunk 5]
```diff
@@ -54,22 +55,38 @@ public CacheMediator createNode(IGraphicalEditPart part, AbstractMediator object
...
-				executeSetValueCommand(CACHE_MEDIATOR__CACHE_SCOPE, CacheScope.PER_MEDIATOR);
-			} else if ("per-host".equals(mediator.getScope())) {
-				executeSetValueCommand(CACHE_MEDIATOR__CACHE_SCOPE, CacheScope.PER_HOST);
-			}
 
 			if (mediator.isCollector()) {
 				executeSetValueCommand(CACHE_MEDIATOR__CACHE_ACTION, CacheAction.COLLECTOR);
 			} else {
+				String[] methods = mediator.getHTTPMethodsToCache();
+				if (!(methods.length == 0 && "".equals(methods[0]))) {
```
Comment (milindaperera) [inline plugins/org.wso2.developerstudio.eclipse.gmf.esb.diagram/src/org/wso2/developerstudio/eclipse/gmf/esb/diagram/custom/deserializer/CacheMediatorDeserializer.java]: Possible NPE when methods.length is actually 0. methods[0] actually does not exists

File: plugins/org.wso2.developerstudio.eclipse.gmf.esb/src/org/wso2/developerstudio/eclipse/gmf/esb/CacheProtocolType.java [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+	 * @generated
+	 */
+	public static CacheProtocolType get(String literal) {
+		for (int i = 0; i < VALUES_ARRAY.length; ++i) {
+			CacheProtocolType result = VALUES_ARRAY[i];
+			if (result.toString().equals(literal)) {
+				return result;
+			}
+		}
+		return null;
```
Comment (manoramahp) [inline plugins/org.wso2.developerstudio.eclipse.gmf.esb/src/org/wso2/developerstudio/eclipse/gmf/esb/CacheProtocolType.java]: We can use for each loop here;
for(CacheProtocolType result : VALUES_ARRAY) {
    if (result.toString().equals(literal)) {
        return result;
    }
}

File: plugins/org.wso2.developerstudio.eclipse.gmf.esb/src/org/wso2/developerstudio/eclipse/gmf/esb/CacheProtocolType.java [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+	 * @return the matching enumerator or <code>null</code>.
+	 * @generated
+	 */
+	public static CacheProtocolType get(String literal) {
+		for (int i = 0; i < VALUES_ARRAY.length; ++i) {
+			CacheProtocolType result = VALUES_ARRAY[i];
+			if (result.toString().equals(literal)) {
+				return result;
+			}
+		}
```
Comment (manoramahp) [inline plugins/org.wso2.developerstudio.eclipse.gmf.esb/src/org/wso2/developerstudio/eclipse/gmf/esb/CacheProtocolType.java]: We can use for each loop here;
for(CacheProtocolType result : VALUES_ARRAY) {
if (result.toString().equals(literal)) {
return result;
}
}

File: plugins/org.wso2.developerstudio.eclipse.gmf.esb.diagram/src/org/wso2/developerstudio/eclipse/gmf/esb/diagram/custom/deserializer/CacheMediatorExtFactory.java (hunk 2 above)
Comment (milindaperera) [inline plugins/org.wso2.developerstudio.eclipse.gmf.esb.diagram/src/org/wso2/developerstudio/eclipse/gmf/esb/diagram/custom/deserializer/CacheMediatorExtFactory.java]: What is the logic behind this: (child.getText() != null && child.getText() != null)

File: plugins/org.wso2.developerstudio.eclipse.gmf.esb/src/org/wso2/developerstudio/eclipse/gmf/esb/CacheProtocolType.java [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+	 * @return the matching enumerator or <code>null</code>.
+	 * @generated
+	 */
+	public static CacheProtocolType getByName(String name) {
+		for (int i = 0; i < VALUES_ARRAY.length; ++i) {
+			CacheProtocolType result = VALUES_ARRAY[i];
+			if (result.getName().equals(name)) {
+				return result;
+			}
+		}
```
Comment (manoramahp) [inline plugins/org.wso2.developerstudio.eclipse.gmf.esb/src/org/wso2/developerstudio/eclipse/gmf/esb/CacheProtocolType.java]: similar as the above for loop

File: plugins/org.wso2.developerstudio.eclipse.gmf.esb.persistence/src/org/wso2/developerstudio/eclipse/gmf/esb/internal/persistence/CacheMediatorTransformer.java [hunk 9]
```diff
@@ -68,30 +70,31 @@ public void transformWithinSequence(TransformationInfo information,
...
 		org.wso2.carbon.mediator.cache.CacheMediator cacheMediator = new org.wso2.carbon.mediator.cache.CacheMediator();
 		setCommonProperties(cacheMediator, visualCache);
 		{	
-			if(visualCache.getCacheAction().getValue()==0){
-				if (StringUtils.isNotBlank(visualCache.getCacheId())) {
-					cacheMediator.setId(visualCache.getCacheId());
-				}			
-				cacheMediator.setScope(visualCache.getCacheScope().getLiteral());
-				cacheMediator.setTimeout(visualCache.getCacheTimeout());
+			if(visualCache.getCacheAction().getValue() == 0){
```
Comment (Manuri) [inline plugins/org.wso2.developerstudio.eclipse.gmf.esb.persistence/src/org/wso2/developerstudio/eclipse/gmf/esb/internal/persistence/CacheMediatorTransformer.java]: Fix formatting.
When does this become 0?

File: plugins/org.wso2.developerstudio.eclipse.gmf.esb.persistence/src/org/wso2/developerstudio/eclipse/gmf/esb/internal/persistence/CacheMediatorTransformer.java [hunk 10]
```diff
@@ -68,30 +70,31 @@ public void transformWithinSequence(TransformationInfo information,
...
+				cacheMediator.setResponseCodes(visualCache.getResponseCodes());
 				cacheMediator.setMaxMessageSize(visualCache.getMaxMessageSize());
-				org.wso2.carbon.mediator.cache.digest.DOMHASHGenerator domhashGenerator = new org.wso2.carbon.mediator.cache.digest.DOMHASHGenerator();
-				cacheMediator.setDigestGenerator(domhashGenerator);
+				cacheMediator.setHttpMethod("http");
+				
+				cacheMediator.setTimeout(visualCache.getCacheTimeout());
+				cacheMediator.setHeadersToExcludeInHash(visualCache.getHeadersToExcludeInHash().split(","));
+				
+				if(visualCache.getHashGenerator().equals("HTTP_REQUEST_HASH_GENERATOR")) {
```
Comment (Manuri) [inline plugins/org.wso2.developerstudio.eclipse.gmf.esb.persistence/src/org/wso2/developerstudio/eclipse/gmf/esb/internal/persistence/CacheMediatorTransformer.java]: Fix formatting

File: plugins/org.wso2.developerstudio.eclipse.gmf.esb.diagram/src/org/wso2/developerstudio/eclipse/gmf/esb/diagram/custom/deserializer/CacheMediatorExtFactory.java [hunk 11]
```diff
@@ -123,49 +131,19 @@ public Mediator createSpecificMediator(OMElement elem, Properties properties) {
...
-                            cacheMediator.setDiskCacheSize(Integer.parseInt(sizeAttr.getAttributeValue()));
-                        } else {
-                            cacheMediator.setDiskCacheSize(DEFAULT_DISK_CACHE_SIZE);
-                        }
-                    } else {
-                        handleException("unknown implementation type for the Cache mediator");
-                    }
-                }
-            }
+			for (Iterator itr = elem.getChildrenWithName(IMPLEMENTATION_Q); itr.hasNext();) {
```
Comment (Manuri) [inline plugins/org.wso2.developerstudio.eclipse.gmf.esb.diagram/src/org/wso2/developerstudio/eclipse/gmf/esb/diagram/custom/deserializer/CacheMediatorExtFactory.java]: Is indentation correct here?

File: plugins/org.wso2.developerstudio.eclipse.gmf.esb/src/org/wso2/developerstudio/eclipse/gmf/esb/CacheProtocolType.java (hunk 7 above)
Comment (prabushi) [inline plugins/org.wso2.developerstudio.eclipse.gmf.esb/src/org/wso2/developerstudio/eclipse/gmf/esb/CacheProtocolType.java]: This is an auto generated code from gmf model. Cannot change it.

File: plugins/org.wso2.developerstudio.eclipse.gmf.esb/src/org/wso2/developerstudio/eclipse/gmf/esb/CacheProtocolType.java (hunk 8 above)
Comment (prabushi) [inline plugins/org.wso2.developerstudio.eclipse.gmf.esb/src/org/wso2/developerstudio/eclipse/gmf/esb/CacheProtocolType.java]: This is an auto generated code from gmf model. Cannot change it.

File: plugins/org.wso2.developerstudio.eclipse.gmf.esb.persistence/src/org/wso2/developerstudio/eclipse/gmf/esb/internal/persistence/CacheMediatorTransformer.java (hunk 9 above)
Comment (prabushi) [inline plugins/org.wso2.developerstudio.eclipse.gmf.esb.persistence/src/org/wso2/developerstudio/eclipse/gmf/esb/internal/persistence/CacheMediatorTransformer.java]: getValue method returns the value generated by the gmf model for that particular cache action. According to the generated gmf model, 0 represents the "FINDER" action. 
will check the formatting.

File: plugins/org.wso2.developerstudio.eclipse.gmf.esb.diagram/src/org/wso2/developerstudio/eclipse/gmf/esb/diagram/custom/deserializer/CacheMediatorDeserializer.java [hunk 12]
```diff
@@ -54,22 +55,38 @@ public CacheMediator createNode(IGraphicalEditPart part, AbstractMediator object
...
-				executeSetValueCommand(CACHE_MEDIATOR__CACHE_SCOPE, CacheScope.PER_MEDIATOR);
-			} else if ("per-host".equals(mediator.getScope())) {
-				executeSetValueCommand(CACHE_MEDIATOR__CACHE_SCOPE, CacheScope.PER_HOST);
-			}
 
 			if (mediator.isCollector()) {
 				executeSetValueCommand(CACHE_MEDIATOR__CACHE_ACTION, CacheAction.COLLECTOR);
 			} else {
+				String[] methods = mediator.getHTTPMethodsToCache();
+				if (methods.length != 0 && !methods[0].isEmpty()) {
```
Comment (sdkottegoda) [inline plugins/org.wso2.developerstudio.eclipse.gmf.esb.diagram/src/org/wso2/developerstudio/eclipse/gmf/esb/diagram/custom/deserializer/CacheMediatorDeserializer.java]: Why do we check if methods[0] is empty?

File: plugins/org.wso2.developerstudio.eclipse.gmf.esb.diagram/src/org/wso2/developerstudio/eclipse/gmf/esb/diagram/custom/deserializer/CacheMediatorDeserializer.java [hunk 13]
```diff
@@ -54,22 +55,38 @@ public CacheMediator createNode(IGraphicalEditPart part, AbstractMediator object
...
+					StringBuilder method = new StringBuilder();
+					for (int i = 0; i < methods.length; i++) {
```
Comment (sdkottegoda) [inline plugins/org.wso2.developerstudio.eclipse.gmf.esb.diagram/src/org/wso2/developerstudio/eclipse/gmf/esb/diagram/custom/deserializer/CacheMediatorDeserializer.java]: We can simply have for (int i = 0; i < methods.length - 1; i++)  and add the code inside else after the for loop. WDYT?

File: plugins/org.wso2.developerstudio.eclipse.gmf.esb.diagram/src/org/wso2/developerstudio/eclipse/gmf/esb/diagram/custom/deserializer/CacheMediatorDeserializer.java (hunk 12 above)
Comment (prabushi) [inline plugins/org.wso2.developerstudio.eclipse.gmf.esb.diagram/src/org/wso2/developerstudio/eclipse/gmf/esb/diagram/custom/deserializer/CacheMediatorDeserializer.java]: This is the logic we have added from the carbon-mediation side. Added the same for dev-studio as well.

File: plugins/org.wso2.developerstudio.eclipse.gmf.esb.diagram/src/org/wso2/developerstudio/eclipse/gmf/esb/diagram/custom/deserializer/CacheMediatorDeserializer.java [hunk 14]
```diff
@@ -54,22 +55,38 @@ public CacheMediator createNode(IGraphicalEditPart part, AbstractMediator object
...
+					}
+					executeSetValueCommand(CACHE_MEDIATOR__CACHE_PROTOCOL_METHODS, method.toString());
+				}
+				executeSetValueCommand(CACHE_MEDIATOR__CACHE_PROTOCOL_TYPE, mediator.getProtocolType());
+				executeSetValueCommand(CACHE_MEDIATOR__RESPONSE_CODES, mediator.getResponseCodes());
+				executeSetValueCommand(CACHE_MEDIATOR__MAX_MESSAGE_SIZE, mediator.getMaxMessageSize());
+
+				String[] headerstoExclude = mediator.getHeadersToExcludeInHash();
+				StringBuilder header = new StringBuilder();
+				for (int i = 0; i < headerstoExclude.length; i++) {
```
Comment (sdkottegoda) [inline plugins/org.wso2.developerstudio.eclipse.gmf.esb.diagram/src/org/wso2/developerstudio/eclipse/gmf/esb/diagram/custom/deserializer/CacheMediatorDeserializer.java]: Same as above. Do we need the if inside the for loop? We can simply loop until headerstoExclude.length and add the code in else after the loop.

File: plugins/org.wso2.developerstudio.eclipse.gmf.esb.diagram/src/org/wso2/developerstudio/eclipse/gmf/esb/diagram/custom/deserializer/CacheMediatorExtFactory.java [hunk 15]
```diff
@@ -23,149 +23,127 @@
...
-						cacheMediator.setDigestGenerator((DigestGenerator) generatorOject);
-					} else {
-						handleException("Specified class for the hashGenerator is not a "
-								+ "DigestGenerator. It *must* implement "
-								+ "org.wso2.carbon.mediator.cache.digest.DigestGenerator interface");
+	public Mediator createSpecificMediator(OMElement elem, Properties properties) {
+
+		if (!CACHE_Q.equals(elem.getQName())) {
+			handleException(
+					"Unable to create the cache mediator. " + "Unexpected element as the cache mediator configuration");
```
Comment (sdkottegoda) [inline plugins/org.wso2.developerstudio.eclipse.gmf.esb.diagram/src/org/wso2/developerstudio/eclipse/gmf/esb/diagram/custom/deserializer/CacheMediatorExtFactory.java]: There is a formatting issue. Shall we fix it?

File: plugins/org.wso2.developerstudio.eclipse.gmf.esb.diagram/src/org/wso2/developerstudio/eclipse/gmf/esb/diagram/custom/deserializer/CacheMediatorExtFactory.java [hunk 16]
```diff
@@ -23,149 +23,127 @@
...
+					OMElement child = (OMElement) itrChildren.next();
+
+					if (child.getText() != null && !"".equals(child.getText())) {
+						try {
+							Class generator = Class.forName(child.getText());
+							Object generatorOject = generator.newInstance();
+							if (generatorOject instanceof DigestGenerator) {
+								cacheMediator.setDigestGenerator((DigestGenerator) generatorOject);
+							} else {
+								handleException("Specified class for the hashGenerator is not a "
```
Comment (sdkottegoda) [inline plugins/org.wso2.developerstudio.eclipse.gmf.esb.diagram/src/org/wso2/developerstudio/eclipse/gmf/esb/diagram/custom/deserializer/CacheMediatorExtFactory.java]: There seems to be a formatting iuuse

**Comment by prabushi** [comment]
Moved changes to https://github.com/wso2/devstudio-tooling-esb/pull/238

**Pull request closed without merging**